### DISCUSSION TITLE
Data modeling for many to many?

### DISCUSSION THREAD

**Discussion by pierroo** [supabase/supabase#16234] [Questions] [answered]
I am quite new to postgres & supabase.
I am trying to do a "many to many" relationship, between "teams / group chats" and their "members"

I found this discussion: https://github.com/orgs/supabase/discussions/710
That mentions "a many-to-many relationship is created by essentially creating 2 tables, then a final "join" table to connect these together."

So in my case I have the tables:
- chats (teams), with id only
- chats_users (id / chat_id => fk on chats.id / user_id => fk on users.id)
- users with all users info

the "chats" are shared by two users (1:1 discussion)

But I really don't see how I can build the query with the join in one trip:
- select the list of chats **I belong to**, as well as the other user info for each of those chats

Can anyone help please? :/

**Comment by GaryAustin1** [comment]
You would do `.from('chats').select('*,users(*)').eq('id',chatID)` or something similar.
As long as you make the two fk's in chat_users both part of the primary key, then the API will figure out the connection.

**Reply by GaryAustin1** [reply, marked as answer]
Edit: the above gets you a specific chat and all the users.  To get your chats then you would do something like:
`.from('chats').select('*, allusers:users(*), users!inner(id)').eq('users.id',userID)`
Note the users table is used twice as you want to get all users and also filter on users.

https://github.com/orgs/supabase/discussions/8385

**Reply by pierroo** [reply]
Thank you @GaryAustin1 for looking into it and providing your insight!
your query + the link were super helpful.

I can confirm that once I changed my "chats_users" table to have both chat_id AND user_id as primary keys, the request you provided work perfectly!

I used the following to clarify:
`.select('*, myself_join:users(id, name), all_users:users!inner(id, name)')`

Although I am wondering if there is anyway in "all_users" to only return the other user (the friend), instead of an array containing both me and my friends; or if I should just filter and clean it from the client and move on with it.

====

Now if I may push my luck,
considering a list of friends like in any discussion app (messenger etc)
=> when a new message is pushed, the chats goes to the top;
when a new conversation is created, it goes to the top etc

How would it work with realtime listeners here, where I would be listening to only the chats I belong to?
(the RLS that goes with it will be a next step I guess)

Should I just listen to the chats.id eq those fetched from initial query above?
and have a "last_msg" column updated for new message to bring it to the top?

That would mean the associated user info are not part of the listener, but I guess it can just be re-fetched from the client, although it doesn't feel ideal to me.

Thank you once again if you happen to pass by to further help! :)

**Reply by GaryAustin1** [reply]
I don't know how to keep the original user out of the query set in the query.  You could certainly do a post process in the client as you know the user id.   I have a simple chat feature in my app and just show all users in the room including the current user.

Realtime will obey RLS so you as long as you have it only allow users who have an id pointing to it then you are safe on chats. You might be able to use the list of chat room ids for the user with the realtime in filter to then just get there room updates. 

If you stored all the users in the chat already then you could possibly use that data as a "cache" to avoid a call back to the server to get the user details if you get a hit.  You would probably need to keep your user list updated though by also having realtime on your chats_users table so you capture users joining or leaving.  It gets complicated fast.

**Reply by pierroo** [reply]
"it gets complicated fast" sums it up perfectly.

Sometimes I wish apps like discord etc would disclose in details how they deal with this whole realtime on chats / groups / presence system without it being super intense for both the client and the server

**Reply by GaryAustin1** [reply]
I think you could just cache the userinfo, maybe when the chat window opens.  Then hit that as a message comes in.  If it misses then fetch that user.  No realtime on it.  Then your risk is someone changes their username or avatar while you have cached info.  Username is not worth the effort for that small window IMO and if your avatars (assuming you are using them) are based on a user id pathname instead of a user supplied pathname  then those won't change either.  Not sure what other user info you might need.  

Technically if you are down to those two you probably could get away with username and id (also used for avatar path) in the actual message columns and not need user info table at all.  In this case older messages might have an old username if they change it.

**Reply by pierroo** [reply]
I fully understand; this way of duplicating data is something I got used to do all these years with nosql;
But I also decided to move to supabase / postgres since I was really tired of having to monitor / update / track all these duplicated data in so many nodes (using firebase for many, many years)

HOWEVER you mentioned a super interesting point here: "if your avatars are based on a user id pathname"
=> this is also something I could NOT do because of how firebase storage work and their image cannot be called through a direct url with just a user_id parameter supplied (contains many tokens and other information that would be changed upon changing avatar photo in this example)

Is supabase any different in that?
Or do you use a specific media storage for these user id pathname avatar?
if supabase is actually built in the way you describe it I would be truly happy, gotta investigate it now :)

EDIT: apparently they do have it, I just have to make the bucket public and I can have a very simple url to an image! AWESOME!

**Comment by pierroo** [comment]
I come back with further questions on this, if you feel like pushing the discussion @GaryAustin1 ? :)

I am now trying to "listen" to my "chats list".
That means listening to
update ("last_msg" field of chats has changed),
delete (chat was deleted, gotta remove it from my list)
insert (I got a new friend / chat)

The problem is that I see two different needs here
- listening to chat UPDATES (mostly on the "last_msg" part, or the "updated_at" so I can order the list accordingly from front) can only happen by listening to "chats" table (shared by both users)
- but INSERT and DELETE can only be listened from chats_users table since this is the one that contains my own user_id needed to filter (only MY chats)

And also, the listening on "chats" table must be filtered on some value, but which one since it doesn't contain reference to my own user_id?
Can we use the same join as the previous select query in a listener? (I don't think so)

Or perhaps I should just duplicate the "last_msg" part, and have it part of the "chats_users"; so I only have to listen to this table?
But this feels very "nosql" thing to do; and what if later I want these chats to be group chats with dozens of people? I don't want to end up duplicating data into each chats_users row for each user when it should only be centralized in the chats table?

is there no easier way to have all the listeners inside one query / channel only?

Somehow ideally I would still love to have just the "chats" table, get rid of "chats_users", and have a "user_ids" array of some sort, containing the two members of that chat, and query on that? (if my user_id is part of this array)
But not sure this is how it should be done with postgres?

I hope you already managed this issue with your chat app, your help would be truly appreciated :)

**Reply by GaryAustin1** [reply]
Realtime can't do a filter on an array column currently.
Although RLS could.
